src/ai/backend/manager/api/vfolder.py

```diff
@@ -316,6 +323,49 @@ async def _wrapped(request: web.Request, *args: P.args, **kwargs: P.kwargs) -> w
     return _wrapped
 
 
+@web.middleware
+async def audit_log_middleware(request: web.Request, handler: Handler) -> web.StreamResponse:
+    # TODO: in near future we can change this condition to dict so we can add more conditions
+
+    if request.method == "GET":
+        return await handler(request)
+
+    root_ctx: RootContext = request.app["_root.context"]
+    user_uuid = str(request["user"]["uuid"])
+    access_key = request["keypair"]["access_key"]
+    user_email = request["user"]["email"]
+
+    audit_log_data.set({
+        "user_id": user_uuid,
+        "access_key": access_key,
+        "email": user_email,
+        "action": None,
+        "data": {"before": {}, "after": {}},
+        "target_type": None,
+        "target": None,
+        "created_at": datetime.utcnow(),
+        "success": False,
+        "rest_api_path": request.path,
+    })
+
+    try:
+        return await handler(request)
+    except Exception:
+        raise
+    finally:
+        try:
+            log.info("AUDIT_LOG: {}", audit_log_data.get())
+            async with root_ctx.db.begin_session() as sess:
+                await sess.execute(sa.insert(audit_logs).values(audit_log_data.get()))
+        except Exception as e:
+            log.error("Failed to write audit log {}", e, exc_info=True)
+
+
+def updated_data(new_data: dict[str, Any]) -> dict[str, Any]:
+    current_audit_log_data = audit_log_data.get().copy()
+    return deep_update(current_audit_log_data, new_data)
```

Comment:
```suggestion
def update_audit_log_data(new_data: dict[str, Any]) -> None:
    audit_log_data.set(
      {**audit_log_data.get(), **new_data}
    )
```

This will work.
And let's not use pydantic v1 API

Reply:
maybe I think using this function is okay
![CleanShot 2024-01-10 at 14 40 32@2x](https://github.com/lablup/backend.ai/assets/37329424/8e17da79-d92c-4e79-9b9b-6223c6212345)

https://docs.pydantic.dev/latest/migration/

this documents says it  `pydantic.v1.utils import deep_update` does not mean it has been deprecated!